enhancement(policy, aws): Construct CIDR rules for referenced AWS security groups by querying associated network interfaces

spacepants opened the pull request:
👋 Hello!

In the current implementation, Cilium can enforce policies based on AWS metadata by cross-walking referenced security group IDs or names to the IP addresses of EC2 instances associated with matching security groups. This PR extends this functionality to any VPC-enabled service, e.g. RDS, ElastiCache, VPC endpoints, etc. by querying the underlying network interfaces directly.

```release-note
Extend AWS metadata-based policy enforcement to work with any VPC-enabled service.
```

First time contributor. Please let me know if this need anything else. Thanks!

christarazi commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -32,39 +32,90 @@ func GetIPsFromGroup(ctx context.Context, group *api.ToGroups) ([]netip.Addr, er
+
+	for labelKey, labelValue := range filter.Labels {
+		newFilter := ec2_types.Filter{
+			Name:   aws.String(fmt.Sprintf("%s:%s", policyEC2Labelskey, labelKey)),
```
```suggestion
			Name:   aws.String(policyEC2Labelskey +":"+ labelKey),
```

christarazi reviewed (commented):
Thanks for the PR!

Could you split up the PR into separate commits where the refactoring is separate from the new additions? This will make it easier to review. For example, the newly defined function could be part of a separate prior commit and then the next could use the commit.

zacharysarah commented on Documentation/security/aws.rst:
```diff
@@ -11,9 +11,9 @@ Locking Down External Access Using AWS Metadata
-15-30 minutes with some experience running Kubernetes.
+based on AWS metadata. It is a detailed walk-through of getting a single-node
+Cilium environment running on your machine. It is designed to take 15-30 minutes
+with some experience running Kubernetes.
```
Edits for clarity
```suggestion
based on AWS metadata. It provides a detailed walk-through of running a single-node
Cilium environment on your machine. It is designed to take 15-30 minutes
for users with some experience running Kubernetes.
```

zacharysarah commented on Documentation/security/aws.rst:
```diff
@@ -155,12 +155,12 @@ Configure AWS Security Groups
+In this example, the destination EC2 elastic network interfaces are attached to
+EC2 instances that are members of a single AWS security group ('sg-0f2146100a88d03c3')
+and pods with label class=xwing should only be able to make connections outside the
+cluster to the destination network interfaces in that security group.
```
Use clear, plain language, formatting
```suggestion
In this example, the destination EC2 elastic network interfaces are attached to
EC2 instances that are members of a single AWS security group ('sg-0f2146100a88d03c3').
Pods with label ``class=xwing`` should only be able to make connections outside the
cluster to the destination network interfaces in that security group.
```

zacharysarah commented on Documentation/security/aws.rst:
```diff
@@ -236,17 +236,17 @@ Policy Language:
-This policy allows traffic from pod *xwing* to any AWS instance that is in
-the security group with ID ``sg-0f2146100a88d03c3``.
+This policy allows traffic from pod *xwing* to any AWS elastic network interface
+that is in the security group with ID ``sg-0f2146100a88d03c3``.
```
Edits for clarity
```suggestion
This policy allows traffic from pod *xwing* to any AWS elastic network interface
in the security group with ID ``sg-0f2146100a88d03c3``.
```

zacharysarah commented on Documentation/security/aws.rst:
```diff
@@ -236,17 +236,17 @@ Policy Language:
-all instances that are part of a specified security group. The list of IPs will
-be updated periodically.
+all network interfaces that are part of a specified security group. The list of
+IPs will be updated periodically.
```
Use present tense
```suggestion
IPs is updated periodically.
```

zacharysarah commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -32,39 +32,90 @@ func GetIPsFromGroup(ctx context.Context, group *api.ToGroups) ([]netip.Addr, er
-		newFilter := ec2_types.Filter{
-			Name:   aws.String(fmt.Sprintf("%s:%s", policyEC2Labelskey, labelKey)),
-			Values: []string{labelValue},
+	// If the group has a security group filter, add the ips from the network interfaces
```
Caps nit
```suggestion
	// If the group has a security group filter, add the IPs from the network interfaces
```

zacharysarah commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -32,39 +32,90 @@ func GetIPsFromGroup(ctx context.Context, group *api.ToGroups) ([]netip.Addr, er
+	return result, nil
+}
+
+// getNetworkInterfaceIpsFromFilter returns the ips from the network interfaces for
```
Caps nit
```suggestion
// getNetworkInterfaceIpsFromFilter returns the IPs from the network interfaces for
```

zacharysarah commented on pkg/policy/groups/aws/aws.go:
```diff
@@ -73,13 +124,14 @@ func getInstancesIpsFromFilter(ctx context.Context, filter *api.AWSGroup) ([]net
 	return extractIPs(result), nil
 }
 
+// extractIPs returns the private and associated public ips from the given reservations
```
Caps nit
```suggestion
// extractIPs returns the private and associated public IPs from the given reservations
```

zacharysarah reviewed (changes requested):
@spacepants Welcome to the project and thanks for the contribution. Some nits to fix, otherwise LGTM. 

Be sure to check out the [documentation style guide](https://docs.cilium.io/en/stable/contributing/docs/docsstyle/).

spacepants committed:
feat(policy, aws): Add new function for returning IP info from AWS ENIs

This adds a new getNetworkInterfaceIpsFromFilter function that returns the associated IP address info for a given query filter.

Signed-off-by: Paul Bailey <spacepants@users.noreply.github.com>

spacepants committed:
refactor(policy, aws): Use ENI query for returning security group IPs

GetIPsFromGroup now calls getNetworkInterfaceIpsFromFilter for all security group related queries as it provides functionally equivalent info for EC2 instance associated ENIs in addition to ENIs associated with other services. EC2 tag related queries still use getInstancesIpsFromFilter. EC2 client construction is moved upstream to GetIPsFromGroup and the client is plumbed into both getNetworkInterfaceIpsFromFilter and getInstancesIpsFromFilter.

Signed-off-by: Paul Bailey <spacepants@users.noreply.github.com>

spacepants committed:
docs(policy, aws): Update docs to reflect new AWS group lookup behavior 

Signed-off-by: Paul Bailey <spacepants@users.noreply.github.com>

zacharysarah reviewed (approved):
LGTM for docs
